jib-core/src/main/java/com/google/cloud/tools/jib/ncache/DefaultCacheStorageWriter.java

```diff
@@ -75,6 +82,15 @@ private static void moveIfDoesNotExist(Path source, Path destination) throws IOE
 
       } catch (FileAlreadyExistsException alsoIgnored) {
         // Same reasoning
+
+      } catch (DirectoryNotEmptyException ex) {
+        // The file system cannot rename directories, so we must resort to copying the directory.
```

Comment:
This happens only on a file system that doesn't support simply renaming a directory with contents (for example, I tried it on a network-mounted virtual file system). The only possible concurrency issue I can see with this is if the layer file is copied first, and the metadata is not copied yet, then another execution will try to rebuild the layer.

Reply:
Filed #1044

Reply:
Just making a note: `DirectoryNotEmptyException` is thrown in our test case because Java can't delete the source directory: https://github.com/frohoff/jdk8u-jdk/blob/master/src/solaris/classes/sun/nio/fs/UnixCopyFile.java#L497

The exception name is a bit misleading at the higher level, since what it is saying at the low level is that, it failed to delete the source directory hence the source directory remains non-empty. (Copy to the target is already done, but because Java couldn't delete the source directory, it does the cleanup to remove the copied target directory.)

BTW, in our case, the source and the target file system providers were identical: the Linux file system.